include/micm/solver/backward_euler.inl

```diff
@@ -0,0 +1,173 @@
+/* Copyright (C) 2023-2024 National Center for Atmospheric Research
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+enum class MicmBackwardEulerErrc
+{
+  SubStepTooSmall = 1,
+};
+
+namespace std
+{
+  template<>
+  struct is_error_condition_enum<MicmBackwardEulerErrc> : true_type
+  {
+  };
+}  // namespace std
+
+namespace
+{
+  class BackwardEulerErrorCategory : public std::error_category
+  {
+   public:
+    const char* name() const noexcept override
+    {
+      return "MICM BackwardEuler";
+    }
+    std::string message(int ev) const override
+    {
+      switch (static_cast<MicmBackwardEulerErrc>(ev))
+      {
+        case MicmBackwardEulerErrc::SubStepTooSmall:
+          return "Substep too small";
+        default: return "Unknown error";
+      }
+    }
+  };
+
+  const BackwardEulerErrorCategory backwardEulerErrorCategory{};
+}  // namespace
+
+inline std::error_code make_error_code(MicmBackwardEulerErrc e)
+{
+  return { static_cast<int>(e), backwardEulerErrorCategory };
+}
+
+namespace micm
+{
+  inline BackwardEuler::BackwardEuler()
+      : processes_()
+  {
+  }
+
+  inline BackwardEuler::BackwardEuler(
+        const System& system,
+        const std::vector<Process>& processes)
+      : processes_()
+  {
+  }
+
+  enum class LastTimeStepCut
+  {
+    None,
+    Half,
+    Tenth
+  };
+
+  inline void BackwardEuler::Solve(double time_step, auto state, auto linear_solver, auto process_set)
+  {
+    // A fully implicit euler implementation is given by the following equation:
+    // y_{n+1} = y_n + H * f(t_{n+1}, y_{n+1})
+    // This is a root finding problem because you need to know y_{n+1} to compute f(t_{n+1}, y_{n+1})
+    // you need to solve the equation y_{n+1} - y_n - H f(t_{n+1}, y_{n+1}) = 0
+    // We will also use the same logic used by cam-chem to determine the time step
+    // That scheme is this: 
+    // Start with H = time_step
+    // if that fails, try H = H/2
+    // if that fails, try H = H/10
+    // if that fails, return whatever the current integration with the current H is
+
+    double H = time_step;
+    double t = 0.0;
+    double tol = 1.0e-6;
+    auto y = state.variables_;
+    auto temp = y;
+    auto forcing = y;
+    auto residual = y;
+    enum LastTimeStepCut last_time_step_cut = LastTimeStepCut::None;
+
+    auto Yn = y;
+    auto Yn1 = y;
+
+    Process::UpdateState(processes_, state);
+
+    bool singular = false;
+
+    while(t < time_step) {
+
+      bool converged = true;
+
+      do {
```

Comment:
I think we are missing the iteration for the Newton method here. See the Fortran implementation at https://github.com/ESCOMP/CAM/blob/cam_development/src/chemistry/pp_trop_strat_mam4_vbs/mo_imp_sol.F90#L259.

Reply:
oh hey we are. Thanks for that